### PR TITLE
A4A: Implement Cancel subscription action component.

### PR DESCRIPTION
**Pull request by jkguidaven**
This PR implements the Cancel subscription button on the Client subscription page.

**Note that currently the submission will not function as we do not have a backend endpoint for it yet. We will have a follow-up PR to point to the correct endpoint later once we have more info on this.**

Screenshot 2024-06-13 at 3 01 14 PM

Related to https://github.com/Automattic/jetpack-genesis/issues/370

## Proposed Changes

* Implemented the CancelSubscriptionAction component that uses a temporary backend hook to simulate a submission. The submission does not work right now, but the processing state should be testable.
* **Additional**: We need to use the correct client endpoint to fetch product information and display the product name properly.

## Testing Instructions

Follow the test instructions from https://github.com/Automattic/wp-calypso/pull/91334

* Confirm that pressing the cancel subscription button will show a dialog.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/a8c-for-agencies/data/client/use-cancel-client-subscription.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { useMutation, UseMutationOptions, UseMutationResult } from '@tanstack/react-query';
+import wpcom from 'calypso/lib/wp';
+import type { APILicense } from 'calypso/state/partner-portal/types';
```
Comment (yashwin) [inline client/a8c-for-agencies/data/client/use-cancel-client-subscription.ts]: Can we use something else here? I don't think it is going to be `APILicense`

File: client/a8c-for-agencies/sections/client/cancel-subscription-confirmation-dialog/index.tsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+
+	const [ isVisible, setIsVisible ] = useState( false );
+
+	const { data: products, isFetching: isFetchingProductInfo } = useFetchClientProducts();
+
+	const { mutate: cancelSubscription, isPending } = useCancelClientSubscription( {
+		onSuccess: () => {
+			onCancelSubscription?.( subscription );
```
Comment (yashwin) [inline client/a8c-for-agencies/sections/client/cancel-subscription-confirmation-dialog/index.tsx:27]: Can we show a success notice when a subscription is canceled, or are we planning to do that in a follow-up PR?

File: client/a8c-for-agencies/sections/client/cancel-subscription-confirmation-dialog/index.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+			</Button>
+
+			<Dialog
+				className="cancel-subscription-confirmation-dialog"
+				isVisible={ isVisible }
+				buttons={ [
+					<Button onClick={ () => setIsVisible( false ) } disabled={ isPending }>
+						{ translate( 'Keep the subscription' ) }
+					</Button>,
+					<Button onClick={ () => onConfirm() } scary primary busy={ isPending }>
```
Comment (yashwin) [inline client/a8c-for-agencies/sections/client/cancel-subscription-confirmation-dialog/index.tsx]: Can we also disable this button when it is `isPending`?

**Review by yashwin — CHANGES_REQUESTED** [review]
Looks good overall. I have left a few comments. Let me know what you think.

File: client/a8c-for-agencies/sections/client/cancel-subscription-confirmation-dialog/index.tsx (hunk 2 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/sections/client/cancel-subscription-confirmation-dialog/index.tsx:27]: Make sense 👍

File: client/a8c-for-agencies/data/client/use-cancel-client-subscription.ts (hunk 1 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/data/client/use-cancel-client-subscription.ts]: Good catch! I've used the `Subscription` type for now while we await more info about the endpoint. 👍

**Comment by jkguidaven** [comment]
@yashwin, thanks for the great review. I've made some changes to address your comments. Can you have another look? 🙏

**Review by yashwin — APPROVED** [review]
Thanks for addressing my comments. LGTM now 🚢

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/14595181

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @jkguidaven for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/14595181).

**Pull request merged**